### PR TITLE
Add tests for GMP code and start overhauling it

### PR DESCRIPTION
**Pull request by fingolfin**
This PR does a couple of things with the GMP integer code in GAP. I
decided to split it into two parts. This is the first, and hopefully
uncontrovesial part, which lays the foundation for future improvements.
Some highlights:

* added extensive test suite for the functions in `src/gmpints.c`,
  giving confidence in the correctness of subsequent changes
* unified error messages
* fixed a bug where raising 0 to a large negative integer produced 0
  instead of an error
* simplfied and cleaned up the code for several functions; `LtInt` and
  `EqInt` even got faster in some cases
* computing remainders modulo a power of 2 now is faster (e.g. for
  checking whether a number is even / odd)
* removed lots of unused stuff, and made some things internal that were
  previously (needlessly) visible to the outside; this will help with
  future improvements
* added resp. exposed AbsInt/ABS_INT and SignInt/SIGN_INT
* added AbsRat/ABS_RAT and SignRat/SIGN_RAT
* switched library to use ABS_RAT and SIGN_RAT for the library operations
  AbsInt and SignInt
* removed the undocumented and normally unrechable "shortbanner" option
  from the GAP library
* made GAP print the GMP version as part of the banner
* documented some details about our integer storage format
* added a DEBUG_GMP macro which enables some paranoia tests

### CODE REVIEW

**Comment by fingolfin** [comment]
It seems quite some code (ab)uses `AbsInt` and `SignInt` with rational arguments.

**Comment by fingolfin** [comment]
So for `AbsInt` we have `AbsoluteValue` as generalization... what to do for `SignInt`? Perhaps `SignValue`? Or perhaps I should just add `AbsRat` and `SignRat`? Thoughts?

File: lib/integer.gi [hunk 1]
```diff
@@ -1278,33 +1278,30 @@ end);
...
 ##
-InstallGlobalFunction( AbsInt, function( n )
-    if 0 <= n  then return  n;
-    else            return -n;
-    fi;
+#InstallGlobalFunction( AbsInt, ABS_INT );
+InstallGlobalFunction( AbsInt, function(n)
+    if IsInt(n) then return ABS_INT(n); fi;
+    if IsRat(n) then return ABS_RAT(n); fi; # for backwards compatibility
+    Error("AbsInt: <n> must be an integer or rational");
```
Comment (fingolfin) [inline lib/integer.gi]: Let's just change ABS_RAT to accept integers, and perform error handling...

File: src/gmpints.c [hunk 2]
```diff
@@ -76,16 +76,13 @@ extern "C" {
...
-#define IS_LARGEINT(obj)       (  ( TNUM_OBJ(obj) == T_INTPOS ) || \
-                                  ( TNUM_OBJ(obj) == T_INTNEG )             )
+#define VAL_LIMB0(obj)         (*ADDR_INT(obj))
+#define SET_VAL_LIMB0(obj,val) do { *ADDR_INT(obj) = val; } while(0)
+#define IS_INTPOS(obj)         (TNUM_OBJ(obj) == T_INTPOS)
+#define IS_INTNEG(obj)         (TNUM_OBJ(obj) == T_INTNEG)
+#define IS_LARGEINT(obj)       (TNUM_OBJ(obj) == T_INTPOS || TNUM_OBJ(obj) == T_INTNEG )
 
 /* for fallbacks to library */
 Obj String;
```
Comment (fingolfin) [inline src/gmpints.c]: Make this static?!?

File: src/gmpints.c [hunk 3]
```diff
@@ -304,10 +289,10 @@ Obj GMP_REDUCE( Obj gmp )
 
 /****************************************************************************
 **
-*F  FuncGMP_INTOBJ(<gmp>) . . . . . . . . . . . . . . . . . . . .  conversion
+*F  GMP_INTOBJ(<gmp>) . . . . . . . . . . . . . . . . . . . .  convert intobj to gmp
 **
 */
-Obj FuncGMP_INTOBJ( Obj self, Obj i )
+Obj GMP_INTOBJ( Obj self, Obj i )
```
Comment (fingolfin) [inline src/gmpints.c]: Get rid of the self parameter

**Comment by fingolfin** [comment]
Note to potential reviewer: I tried to make each commit self contained, and for a *thorough* review, reviewing each commit on its own is probably easiest.

**Comment by fingolfin** [comment]
I also tried to edit the description of this PR so that it can be reused for a merge commit...

**Comment by fingolfin** [comment]
This PR is now stable (i.e. I don't plan to rewrite its commits anymore, unless changes request are made). Reviews are welcome!

File: src/gmpints.c [hunk 4]
```diff
@@ -297,6 +318,38 @@ Obj GMP_REDUCE( Obj gmp )
   return gmp;
 }
 
+#if DEBUG_GMP
```
Comment (fingolfin) [inline src/gmpints.c]: I am not sure about using IS_NORMALIZED_AND_REDUCED from packages (ideally, packages should never assume that they know how GAP integers work internally; I'd prefer to add APIs that do whatever packages need... but of course that just an idealistic wish, and not realistic right now).

In any case, a strong reason why we should indeed always compile `IS_NORMALIZED_AND_REDUCED` is to preven bitrot in it. I'll make that change.

**Comment by fingolfin** [comment]
I thought careful about imposing that requirement on ints always being normalized and reduced. Fact is, a lot of the code already implicitly required that to be correct. In the end, making this the official rule seemed to be the sanest way to cleanup things. Thanks for pointing out the issue with `IntHexString`, I didn't catch it because I did most of my work with an extended version of this branch, in which it as well as all int->string conversion have been completely rewritten. I'll also fix it in this branch, though.

As to the commentend out GMP_REDUCE/GMP_NORMALIZE: these are safe to remove (I had CHECK_GMP calls in there to double check for quite some time). This, too, was not removed on this branch mainly because I completely rewrote the relevant functions. I'll still update it on this branch, too (it's a bit annoying, as it duplicates work, but makes it clearer that the work on this PR is correct).

**Comment by fingolfin** [comment]
Updated the PR based on @ChrisJefferson's feedback; I also added `CHECK_GMP` calls systematically to the start of functions, and also before returns (previously, I only had added them in a few semi-random places)

**Comment by fingolfin** [comment]
@ChrisJefferson @markuspf @frankluebeck I guess this PR is too big and hence does not get any reviews ... :-). While I am reasonably confident that it is good-to-go as it is, perhaps it would help if I split it into a series of more manageble PRs? Alternatively, I point out once again that one can review this commit by commit...

(Or you can not review it and just say "guess it's fine, we trust you to clean up any fall out it may cause" ;-)

**Comment by markuspf** [comment]
I just didn't have the time to look at it yet. It is quite a large PR, but yet it's self contained and fairly uncontroversial. I'll get to it before the end of this week, if that's of any help.

**Comment by fingolfin** [comment]
OK, I'll be patient then. Thanks :)

**Comment by ChrisJefferson** [comment]
I'm happy with this commit, except for the comment I made which you fixed. Not sure why it's now conflicting?

**Comment by fingolfin** [comment]
The conflict is there because you just merged #1056. I'll fix it.

**Comment by markuspf** [comment]
I am happy with this.

**Comment by fingolfin** [comment]
So, merge this? Rebase it? Should I do it myself?

**Comment by ChrisJefferson** [comment]
No-one else has any opinions, so I'll do it.

**Pull request merged**